src/ucp/core/ucp_listener.c

```diff
@@ -20,17 +20,31 @@ static unsigned ucp_listener_conn_request_progress(void *arg)
 
     ucs_trace_func("listener=%p ep=%p", accept->listener, ep);
 
-    /* send wireup request message, to connect the client to the server's new endpoint */
-    status = ucp_wireup_send_request(accept->ep);
-    if (status != UCS_OK) {
-        goto err_destroy_ep;
+    if (!(ep->flags & UCP_EP_FLAG_LISTENER)) {
+        /* send wireup request message, to connect the client to the server's new endpoint */
+        ucs_assert(!(ep->flags & UCP_EP_FLAG_CONNECT_REQ_QUEUED));
+        status = ucp_wireup_send_request(ep);
+        if (status != UCS_OK) {
+            goto err_destroy_ep;
+        }
+    } else {
+        status = ucp_wireup_send_pre_request(ep);
+        if (status != UCS_OK) {
+            goto err_destroy_ep;
+        }
     }
 
     if (accept->listener->cb != NULL) {
-        ep->flags |= UCP_EP_FLAG_USED;
-        accept->listener->cb(ep, accept->listener->arg);
+        if (ep->flags & UCP_EP_FLAG_LISTENER) {
+            ep->flags &= ~UCP_EP_FLAG_USED;
+            ucp_ep_ext_gen(ep)->listener = accept->listener;
```

Comment:
can we use `accept->listener` instead of `ucp_ep_ext_gen(ep)->listener`? so the `if (ep->flags ...` will not be needed?

Reply:
replace where? can you please clarify what you mean? 
here we save the listener for future use. what line to you suggest to replace with what?

Reply:
didn't explain it well, can we keep the "accept" structure until user callback is called, instead of saving the listener on the ep? "accept" is allocated anyway, and also this way we can avoid the union between user_data/listener.
Also, in case of partial address, maybe there is no need to use ucp_listener_conn_request_progress() for the first step, e.g maybe we can call ucp_wireup_send_pre_request(ep) immediately from ucp_listener_conn_request_callback(), since this time no need to call user's accept callback

Reply:
keep the "accept" structure where? the callback will be invoked from the end of wireup flow..

Reply:
@yosefe also, calling ucp_wireup_send_pre_request(ep) immediately from ucp_listener_conn_request_callback() would mean to start the wireup flow from the async thread which we decided not to do a while ago since uct isn't thread safe.